src/ruby/puppetserver-lib/puppet/server/ast_compiler.rb

```diff
@@ -91,34 +111,120 @@ def self.compile_ast(code, compile_options, boltlib_path)
-              # initializes the necessary type loaders for us.
-              ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+            pal.with_catalog_compiler(**variables) do |compiler|
+              # TODO: PUP-10476 Explore setting these as default in PAL. They are the defaults in Puppet
+              Puppet[:strict] = :warning
+              Puppet[:strict_variables] = false
 
+              ast = build_program(code)
               compiler.evaluate(ast)
+              compiler.evaluate_ast_node
               compiler.compile_additions
               compiler.catalog_data_hash
             end
           end
         end
       end
-      private_class_method :compile_ast
+      private_class_method :compile_for_environment_plan
+
+      def self.compile_for_environment(code, compile_options, boltlib_path)
+        # When we do not need to load bolt we assume there are no Bolt types to resolve in
+        # AST compilation.
+        Puppet[:node_name_value] = compile_options['certname']
+
+        # Use the existing environment with the requested name
+        Puppet::Pal.in_environment(compile_options['environment'],
+                                    envpath: Puppet[:environmentpath],
+                                    facts: compile_options['facts']['values'],
+                                    variables: compile_options['variables']['values']) do |pal|
+          Puppet.lookup(:pal_current_node).trusted_data = compile_options['trusted_facts']['values']
+
+          # This compiler has been configured with a node containing
+          # the requested environment, facts, and variables, and is used
+          # to compile a catalog in that context from the supplied AST.
+          pal.with_catalog_compiler do |compiler|
+            # We have to parse the AST inside the compiler block, because it
+            # initializes the necessary type loaders for us.
+            ast = Puppet::Pops::Serialization::FromDataConverter.convert(code)
+
+            compiler.evaluate(ast)
+            compiler.compile_additions
+            compiler.catalog_data_hash
+          end
+        end
+      end
+      private_class_method :compile_for_environment
+
+
+      def self.with_puppet_settings(overrides = {})
+        Dir.mktmpdir('bolt') do |dir|
+          cli = []
+          Puppet::Settings::REQUIRED_APP_SETTINGS.each do |setting|
+            cli << "--#{setting}" << dir
+          end
+          Puppet.settings.send(:clear_everything_for_tests)
+          # Override module locations, Bolt includes vendored modules in its internal modulepath.
+          Puppet.settings.override_default(:basemodulepath, '')
+          Puppet.settings.override_default(:vendormoduledir, '')
+
+          Puppet.initialize_settings(cli)
+          overrides.each do |setting, value|
+            Puppet.settings[setting] = value
+          end
+
+          Puppet.settings[:log_level] = @log_level
```

Comment:
`@log_level` wont be set here. Not going to bother making this dynamic for now because we may or may not even go this route. I would suggest either leaving it unset (and hopefully picking up some default) or hard coding it to something reasonable.